kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoKVTable.java

```diff
@@ -97,28 +95,27 @@ public String name() {
 
   @Override
   public WriterFactory<Bytes, Integer> init(final int kafkaPartition) {
-    metadataRows.computeIfAbsent(kafkaPartition, kp -> {
-      final MetadataDoc metaDoc = metadata.findOneAndUpdate(
-          Filters.eq(MetadataDoc.PARTITION, kafkaPartition),
-          Updates.combine(
-              Updates.setOnInsert(MetadataDoc.PARTITION, kafkaPartition),
-              Updates.setOnInsert(MetadataDoc.OFFSET, NO_COMMITTED_OFFSET),
-              Updates.inc(MetadataDoc.EPOCH, 1) // will set the value to 1 if it doesn't exist
-          ),
-          new FindOneAndUpdateOptions()
-              .upsert(true)
-              .returnDocument(ReturnDocument.AFTER)
-      );
-
-      if (metaDoc == null) {
-        throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
-      }
-
-      LOG.info("Retrieved initial metadata {}", metaDoc);
-
-      return metaDoc;
-    });
 
+    final MetadataDoc metaDoc = metadata.findOneAndUpdate(
+        Filters.eq(MetadataDoc.PARTITION, kafkaPartition),
+        Updates.combine(
+            Updates.setOnInsert(MetadataDoc.ID, idForKafkaPartition(kafkaPartition)),
+            Updates.setOnInsert(MetadataDoc.PARTITION, kafkaPartition),
+            Updates.setOnInsert(MetadataDoc.OFFSET, NO_COMMITTED_OFFSET),
+            Updates.inc(MetadataDoc.EPOCH, 1) // will set the value to 1 if it doesn't exist
+        ),
+        new FindOneAndUpdateOptions()
+            .upsert(true)
+            .returnDocument(ReturnDocument.AFTER)
+    );
+
+    if (metaDoc == null) {
+      throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
+    }
+
+    LOG.info("Retrieved initial metadata {}", metaDoc);
+
+    kafkaPartitionToEpoch.put(kafkaPartition, metaDoc.epoch);
```

Comment:
I think changing this from computeIfAbsent to put fixed a bug when rebalances that re-assigned the node the same pod would not generate a new epoch, and therefore remain fenced.